examples/bits.rs

```diff
@@ -1,98 +1,207 @@
+
+// These are helpers for the turtle's actions.
+
+/// Set the horizontal spacing between successive bits in a row
+const BIT_MARGIN: f64 = BIT_WIDTH / 2.0;
+/// Compute the total width of a bit plus its spacing
+const BIT_BOX: f64 = BIT_WIDTH + BIT_MARGIN;
 
 fn main() {
-    let mut bits = BitVec::with_capacity(17);
-    bits.set(0, true);
-    bits.set(1, true);
-    bits.set(3, true);
-    bits.set(8, true);
-    bits.set(10, true);
-    bits.set(16, true);
-    println!("{}", bits);
-    bits.set(0, false);
-    bits.set(8, false);
-    println!("{}", bits);
-}
+    // This block sets up the turtle to draw bits more or less centered in the
+    // screen. The turtle works by walking horizontally for each bit in a byte,
+    // then backtracking and walking vertically to the next byte.
+    let mut turtle = Turtle::new();
+    // The turtle starts in the center, but we want to move it around before
+    // drawing.
+    turtle.pen_up();
+    // Compute the boundaries of the part of the screen where the turtle will
+    // draw
+    let right_edge = BIT_BOX * 4.0;
+    // Place the turtle according to the length of the text span
+    let top_edge = BIT_HEIGHT * (TEXT.len() as f64 / 2.0);
+    // The turtle starts from the top right of the region,
+    turtle.go_to((right_edge, top_edge));
+    // and walks left
+    turtle.set_heading(180.0);
 
-/// The number of bits in a single element of the bytes Vec in BitVec
-/// Set to 8 because BitVec stores u8 values. If we stored u16, we would use 16 here
-const BITS: usize = 8;
-
-/// An implementation of a vector of bits that can be easily set individually
-/// Note that this is meant to be a teaching implementation and so is not necessarily the most
-/// sophisticated or performant way to do everything. The goal is to make the code understandable
-/// and correct.
-/// See the bit-vec crate and others on crates.io for more advanced implementations.
-#[derive(Debug, Clone, Default)]
-pub struct BitVec {
-    bytes: Vec<u8>,
-}
+    // Render the text as bits of memory.
 
-impl fmt::Display for BitVec {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        // EXERCISE: Only print the bytes up to the length of bytes that have been pushed, not
-        // the entire capacity. (Similar to how Vec prints its len, not its capacity)
-        for byte in &self.bytes {
-            write!(f, "{:0width$b}", byte, width = BITS)?;
-        }
-        Ok(())
-    }
-}
+    // Rust strings can iterate over their individual characters. This block
+    // loops over characters, collecting their start point in the text so that
+    // we can grab the encoded bytes of each one.
+    let mut row_num = 0;
+    for (char_num, (start, codepoint)) in TEXT.char_indices().enumerate() {
+        println!("Character {}: {}", char_num, codepoint);
+        // Each character has a variable width, so we need to find that.
+        let byte_count = codepoint.len_utf8();
```

Comment:
ooh, good idea
![image](https://user-images.githubusercontent.com/5921109/70564743-f7ad8f80-1b4d-11ea-91ed-b92fbcfd901c.png)

Reply:
Hmmm. Given that there are gaps between each bit, the extra gaps don't actually totally help to delineate each character. Maybe reduce the spacing between each row of bits in each character? If that doesn't help, you could even draw a thinner or lighter line between each character. Definitely something to play around with to see how we can indicate separation between each character's bytes.

Reply:
![image](https://user-images.githubusercontent.com/5921109/70567723-b9b36a00-1b53-11ea-85db-c692b5e3f174.png)

Using regular grey for the divider, light grey for off, and black for on, gives pretty decent contrast to me. I've also added comments explaining that bytes in a character are grouped, and after each character is a divider line.

Reply:
Looks good! I'm looking forward to reviewing the revision. :)

Is there a reason why the lines extend all the way to the last gap on the left? Do you think it would look better if the lines stopped at the last visible portion of the bits?

Reply:
because I didn't subtract a `BIT_MARGIN` from the `8 * BIT_BOX` computation when sketching this. It's fixed locally.